module/zfs/zvol.c

```diff
@@ -2239,18 +2236,31 @@ zvol_set_snapdev_sync_cb(dsl_pool_t *dp, dsl_dataset_t *ds, void *arg)
 }
 
 /*
- * Traverse all child snapshot datasets and apply snapdev appropriately.
+ * Traverse all child datasets and apply snapdev appropriately.
+ * We call dsl_prop_set_sync_impl() here to set the value only on the toplevel
```

Comment:
@loli10K, @behlendorf 

I think the snapdev property and how it interacts with the property inheritance can benefit from clarification at this time.

The original code has always propagated the `snapdev` settings for the top-level filesystem through all the children filesystems, as well as all their children zvols.
```
[root@centos-6 zfs-bprotopopov]# for z in $(zfs list -H | cut -f1); do zfs get -H snapdev $z; done
tpool	snapdev	hidden	default
tpool/top-fs	snapdev	visible	local
tpool/top-fs/sub-fs	snapdev	visible	local
tpool/top-fs/sub-fs/zvol0	snapdev	visible	local
tpool/top-fs/sub-fs/zvol1	snapdev	visible	local
tpool/top-fs/zvol0	snapdev	visible	local
tpool/top-fs/zvol1	snapdev	visible	local
tpool/zvol0	snapdev	hidden	local
[root@centos-6 zfs-bprotopopov]# zfs set snapdev=hidden tpool/top-fs
[root@centos-6 zfs-bprotopopov]# for z in $(zfs list -H | cut -f1); do zfs get -H snapdev $z; done
tpool	snapdev	hidden	default
tpool/top-fs	snapdev	hidden	local
tpool/top-fs/sub-fs	snapdev	hidden	local
tpool/top-fs/sub-fs/zvol0	snapdev	hidden	local
tpool/top-fs/sub-fs/zvol1	snapdev	hidden	local
tpool/top-fs/zvol0	snapdev	hidden	local
tpool/top-fs/zvol1	snapdev	hidden	local
tpool/zvol0	snapdev	hidden	local
[root@centos-6 zfs-bprotopopov]#
```
But I believe @loli10K  suggests that this is not quite correct, and that the propagation of the properties needs to be in line with the `zfs inherit` semantics. This seems reasonable to me.

As I understand it, `zfs inherit` command explicitly passes the values of the properties to a specific child or recursively to all the children, but it recurses in the user space (`zfs_for_each()`), so that kernel methods do not need to be concerned with this. If that's the case, then I believe there is an opportunity here to cleanup the code a bit: we should set the value for the dataset in question and then we should only be recursing over snapshots if the dataset is a zvol.

I don't think we need to recurse through the hierarchy and re-read and re-set snapdev for all the children. Also, before setting and recursing, let's check that the value to be set is different from the one that is already set. This can save quite a bit of I/O.

@loli10K has previously brought up a 'hypothetical' use case of changing the setting of `zvol_inhibit_dev` along with setting `snapdev` and suggested that 're-reading and re-setting' would make that use case work in a more intuitive way. I am not convinced this is an important use case to support cleanly (there is a simple workaround of toggling `snapdev` that also works). In my experience, setting zvol_inhibit_dev is a design-time decision, not a runtime one.

Thoughts ?